### PR TITLE
Keep a # delimiter when raw string content contains a quote

### LINKED ISSUE #2465: Refactoring "Convert string literal to minimal number of "#"s" produces incorrect source code
### Swift version

6.2.3

### Platform

Arch Linux

### Editor

Visual Studio Code Swift plugin  2.14.2 and also my own JetBrains plugin (Noctule)

### Description

Refactoring "Convert string literal to minimal number of "#"s" does not handle string literals correctly, which would have to have escaped characters after the conversion.

### Steps to Reproduce

Place the text cursor into the string literal and then invoke the refactoring "Convert string literal to minimal number of "#"s"

```swift
let x = #"""#
```

The file is then turned into this, which is incorrect because the previously literal double quote `"' is now unescaped:
```swift
let x = """
```

I think either the refactoring should not be offered or the result should be properly escaped.

### Logging

_No response_

### PR DESCRIPTION
**Pull request by itsybitsybootsy**
## Problem

`FormatRawStringLiteral` ("Convert string literal to minimal number of `#`s")
strips all `#` delimiters from a single-line raw string when the content has
no `#` characters. That is incorrect when the content contains a `"`:
stripping the delimiters fuses the content quote with the surrounding quotes,
producing either a multi-line string opener (`#"""#` -> `"""`) or an
early-terminated literal (`#"foo"bar"#` -> `"foo"bar"`, which parses as
string `"foo"` followed by identifier `bar` and an unterminated `"`).

The original report on sourcekit-lsp#2465 hinted at a special case for
`#"""#`, but the broken-output property holds for any single-line raw
literal whose content contains a `"`. The fix handles all of those uniformly.

## Fix

In the `maximumHashes == 0` branch, also check whether any string segment
contains a `"`. When it does, and the literal is a single-line raw string
without parse errors, retain a single `#` delimiter instead of stripping
all of them.

Multi-line literals are unchanged (a single `"` in their content does not
fuse with the `"""` delimiters), and literals that already contain a parse
error fall through to the pre-existing behaviour to avoid reasoning about
parser-recovery content.

## Testing

- Added five regression cases in `Tests/SwiftRefactorTest/FormatRawStringLiteral.swift`:
  the original `#"""#` reproducer, mid-content quotes (`#"foo"bar"#`,
  `#"a"b"c"#`), an over-delimited input that reduces correctly
  (`##"hello""world"##` -> `#"hello""world"#`), and a raw string containing
  `\"` (two literal characters in raw mode).
- `SKIP_LONG_TESTS=1 swift test`: 3531 passed, 8 skipped, 0 failed (same as `main`).

Resolves swiftlang/sourcekit-lsp#2465.

---

This contribution was developed with the help of Claude Code. The
implementation, regression tests, edge-case analysis, and the local test
suite were reviewed manually before submitting.

### CODE REVIEW

**Comment by ahoppen** [comment]
Thanks for the PR. There was a previous attempt to fix this: https://github.com/swiftlang/swift-syntax/pull/3256. Could you ensure that all the comments that were raised in that PR are also addressed in yours?

**Comment by itsybitsybootsy** [comment]
Read through #3256, thanks. My change is the narrow case, doesn't cover the backslash escape variants or `\###n` cases that parse-and-verify would catch.

Want me to rewrite this to use parse-and-verify, or close it for #3256? itsdevandy has been quiet for three months.

**Comment by ahoppen** [comment]
This PR should also take care of the backslash variants since we’ve identified them as issues and they are very related to what you’re fixing.

And since https://github.com/swiftlang/swift-syntax/pull/3256 hasn’t had any activity for a while, you can take over the issue.

**Comment by itsybitsybootsy** [comment]
Rewritten to parse-and-verify and pushed. Tests cover the four cases: `#"""#`, internal quotes, `\` escapes, and `\#n`.

File: Sources/SwiftRefactor/FormatRawStringLiteral.swift [hunk 1]
```diff
@@ -54,14 +60,46 @@ public struct FormatRawStringLiteral: SyntaxRefactoringProvider {
...
-        .with(\.openingPounds, lit.openingPounds?.with(\.tokenKind, .rawStringPoundDelimiter("")))
-        .with(\.closingPounds, lit.closingPounds?.with(\.tokenKind, .rawStringPoundDelimiter("")))
+    // Search for the smallest delimiter count that preserves the meaning of
+    // the literal. Removing pounds can flip a literal `\#(...)` into an
+    // interpolation, change `\n` from literal to a newline, fuse content
+    // quotes with the surrounding quote token, or turn `"""` into a
+    // multi-line opener. For each candidate count we rebuild the literal,
+    // re-parse it, and accept the count only if (a) the candidate has no
+    // parse errors and (b) it has the same `representedLiteralValue` as the
+    // original. The upper bound `maximumHashes + 1` matches what the older
```
Comment (ahoppen) [inline Sources/SwiftRefactor/FormatRawStringLiteral.swift]: We shouldn’t reference an “older” implementation. A year from now nobody will know what the old implementation was 😉 

Same below

File: Sources/SwiftRefactor/FormatRawStringLiteral.swift [hunk 2]
```diff
@@ -36,6 +38,10 @@ import SwiftSyntax
 /// ```
 public struct FormatRawStringLiteral: SyntaxRefactoringProvider {
   public static func refactor(syntax lit: StringLiteralExprSyntax, in context: Void) -> StringLiteralExprSyntax {
+    // The highest number of consecutive `#` characters that appears in any
+    // string-segment content or interpolation delimiter. Using more than this
+    // many delimiters guarantees a syntactically valid literal, so it gives
+    // us an upper bound for the search below.
     var maximumHashes = 0
```
Comment (ahoppen) [inline Sources/SwiftRefactor/FormatRawStringLiteral.swift]: Do we actually still need this logic? Might it be simpler if we just do the loop below until the current number of `#` characters? Parsing should be fast enough that this isn’t a performance problem.

File: Sources/SwiftRefactor/FormatRawStringLiteral.swift (hunk 1 above)
Comment (itsybitsybootsy) [reply Sources/SwiftRefactor/FormatRawStringLiteral.swift]: Removed those references in the rewrite.

File: Sources/SwiftRefactor/FormatRawStringLiteral.swift (hunk 2 above)
Comment (itsybitsybootsy) [reply Sources/SwiftRefactor/FormatRawStringLiteral.swift]: Applied to the main search path: upper bound is now `lit.openingPounds?.text.count ?? 0`. Kept the content scan localized to the `hasError` branch because the 3 malformed-input tests (lines 24, 28, 30) rely on the content-safe fallback when `representedLiteralValue` can't be used for verification.

Comment (ahoppen) [reply Sources/SwiftRefactor/FormatRawStringLiteral.swift]: I don’t think we should have a totally separate code path for if the string literal has an error. I didn’t think of the error case here. Let’s use your previous implementation then.

Comment (itsybitsybootsy) [reply Sources/SwiftRefactor/FormatRawStringLiteral.swift]: Reverted to the unified path. Tests pass.

File: Sources/SwiftRefactor/FormatRawStringLiteral.swift [hunk 3]
```diff
@@ -36,16 +38,19 @@ import SwiftSyntax
...
+    // string-segment content or interpolation delimiter. Using one more than
+    // this many delimiters always produces a syntactically valid literal, so
+    // it serves both as an upper bound for the search below and as a safe
+    // count to fall back to when the input is malformed.
     var maximumHashes = 0
     for segment in lit.segments {
       switch segment {
       case .expressionSegment(let expr):
         if let rawStringDelimiter = expr.pounds {
-          // Pick up any delimiters in interpolation segments \#...#(...)
```
Comment (ahoppen) [inline Sources/SwiftRefactor/FormatRawStringLiteral.swift:44]: Did you intentionally remove these comments?

Since you are using an LLM to assist you in this PR, please always review the changes it made according to our [AI Policy](https://github.com/swiftlang/project-operations/blob/main/contributing/ai-tools.md). Since you are fixing an issue marked *good first issue* it is explicitly discourage to over-rely on AI tools, which I believe may be happening here. If this PR was mostly generated by an LLM, I ask you to close it so this good first issue is left as a learning opportunity to fellow contributors.

**Comment by itsybitsybootsy** [comment]
Fair point on the AI policy. Closing this so the issue stays available as a learning opportunity for another contributor.
